Mathlib/Analysis/InnerProductSpace/Spectrum.lean

```diff
@@ -76,6 +77,31 @@ theorem conj_eigenvalue_eq_self (hT : T.IsSymmetric) {μ : 𝕜} (hμ : HasEigen
   rw [mem_eigenspace_iff] at hv₁
   simpa [hv₂, inner_smul_left, inner_smul_right, hv₁] using hT v v
 
+/-- The generalized eigenspaces of a symmetric operator coincide with the eigenspace. -/
+theorem genEigenspace_eq_eigenspace
+    {T : E →ₗ[𝕜] E} (hT : T.IsSymmetric) {n : ℕ} {μ : 𝕜} (hn : 1 ≤ n) :
+    genEigenspace T μ n = genEigenspace T μ 1 := by
+  refine Nat.le_induction rfl (fun k hk ih ↦ ?_) n hn
+  refine ih ▸ le_antisymm (fun x hx ↦ ?_) ((genEigenspace T μ).mono k.le_succ)
+  obtain (rfl | hx_ne) := eq_or_ne x 0
+  · exact zero_mem _
+  · have hμ : starRingEnd 𝕜 μ = μ := hT.conj_eigenvalue_eq_self <|
+    hasEigenvalue_of_hasGenEigenvalue (k := k + 1) <|
+    (genEigenspace T μ (k + 1)).ne_bot_iff.mpr ⟨x, hx, hx_ne⟩
+    have hTμ : ((T - μ • 1) ^ k).IsSymmetric := by
+      aesop (add simp LinearMap.one_eq_id)
+    rw [mem_genEigenspace, ← norm_eq_zero, ← sq_eq_zero_iff, norm_sq_eq_inner (𝕜 := 𝕜),
+      hTμ, ← LinearMap.comp_apply, ← LinearMap.mul_eq_comp, ← pow_add]
+    simp [mem_genEigenspace .. |>.mp <| (genEigenspace T μ).mono (show k + 1 ≤ k + k by gcongr) hx]
+
+lemma maxGenEigenspace_eq_eigenspace {T : E →ₗ[𝕜] E} (hT : T.IsSymmetric)
+    {μ : 𝕜} : maxGenEigenspace T μ = eigenspace T μ := calc
+  _ = ⨆ n, genEigenspace T μ (n + 1) := by
+    rw [maxGenEigenspace_def, ← sup_iSup_nat_succ, genEigenspace_def]; simp [LinearMap.one_eq_id]
+  _ = ⨆ _ : ℕ, genEigenspace T μ 1 := by
+    congr! 2 with n; exact genEigenspace_eq_eigenspace hT n.succ_pos
+  _ = eigenspace T μ := by simp [genEigenspace_def, eigenspace_def]
+
```

Comment:
The point of Oliver's previous comment:

> Bearing in mind [Module.End.IsSemisimple.genEigenspace_eq_eigenspace](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Eigenspace/Semisimple.html#Module.End.IsSemisimple.genEigenspace_eq_eigenspace) and [Module.End.IsSemisimple.maxGenEigenspace_eq_eigenspace](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Eigenspace/Semisimple.html#Module.End.IsSemisimple.maxGenEigenspace_eq_eigenspace) I wonder if instead of these we could prove ...

is that we don't need these anymore given the proof of `LinearMap.IsSymmetric.isSemisimple` that you gave elsewhere.
```suggestion
```

Reply:
Yes, but these don't require the finite dimensional assumption, so I thought they might be useful elsewhere later because they are hence more general. (I figured the idea was to use as general results as possible for as long as possible. Please disabuse me of this notion if it isn't right!)

Reply:
IMHO the right solution for me to complete #17689 where I introduce the "semisimple on finite-dimensional subspaces" property (which holds for symmetric infinite-dimensional operators) and where I also prove that for any operators satisfying this condition, all generalised eigenvectors are true eigenvectors.

To avoid being blocked by my work, I suggest you follow Jireh's advice and do not add these lemmas, but retain the fact that we have semisimplicity in finite dimensions. I can generalise this to finitely semisimple in all dimensions later on.